File: Leetcode_Code/199. Binary Tree Right Side View/199.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        rightSideView_dfs(root, 0);
        return result;
    }

    private List<Integer> result = new ArrayList<>();

    private void rightSideView_dfs(TreeNode root, int layer) {
        if (root == null) return;
        if (result.size() <= layer) result.add(root.val);
        rightSideView_dfs(root.right, layer + 1);
        rightSideView_dfs(root.left, layer + 1);
    }
}

File: Leetcode_Code/199. Binary Tree Right Side View/199.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> result;
        rightSideView_dfs(root, 0, result);
        return result;
    }

private:
    void rightSideView_dfs(TreeNode *root, int layer, vector<int> &result) {
        if (!root) return;
        if (result.size() <= layer) result.push_back(root->val);
        rightSideView_dfs(root->right, layer + 1, result);
        rightSideView_dfs(root->left, layer + 1, result);
    }
};
